pyccel/stdlib/ndarrays/ndarrays.c

```diff
@@ -0,0 +1,124 @@
+#include "ndarrays.h"
+
+/* 
+** allocation
+*/
+
+t_ndarray *init_array(char *temp, int nd, int *shape, int type)
+{
+	t_ndarray *a;
+		
+	a = malloc(sizeof(t_ndarray));
+	a->type = type;
+	a->nd = nd;
+	a->shape = malloc(a->nd * sizeof(int));
+	a->lenght = 1;
+	a->is_slice = 0;
+	for (int i = 0; i < a->nd; i++) // init the shapes
+	{
+		a->lenght *= shape[i]; 
+		a->shape[i] = shape[i];
+	}
+	a->strides = malloc(nd * sizeof(int));
+	for (int i = 0; i < a->nd; i++)
+	{
+		a->strides[i] = 1;
+		for (int j = i + 1; j < a->nd; j++)
+			a->strides[i] *= a->shape[j];
+	}
+	a->data = malloc(sizeof(t_ndarray_type));
+	a->data->raw_data = calloc(a->lenght , a->type);
+	if (temp)
+		memcpy(a->data->raw_data, temp, a->lenght * a->type);
+	return (a);
+}
+
+/* 
+** deallocation
+*/
+
+int free_array(t_ndarray *dump)
```

Comment:
If everything is going to be a pointer then I wonder if it would be better to pass a pointer to a pointer. That way you can set the pointer to NULL.
Ie:
```
int free_array(t_ndarray **dump)
{
	if (!(*dump)->is_slice)
	{
		free((*dump)->data->raw_data);
		(*dump)->data->raw_data = NULL;
	}
	free((*dump)->data);
	(*dump)->data = NULL;
	free((*dump)->shape);
	(*dump)->shape = NULL;
	free((*dump)->strides);
	(*dump)->strides = NULL;
	free(*dump);
	*dump = NULL;
	return (1);
}
```